internal/installercache/installercache.go

```diff
@@ -41,53 +48,103 @@ func (fi *fileInfo) Compare(other *fileInfo) bool {
+	// clusterID is the UUID of the cluster for which the release is being fetched.
+	clusterID strfmt.UUID
+	// releaseId is the release that is being fetched, for example "4.10.67-x86_64".
+	releaseID string
+	// cached is `true` if the release was found in the cache, otherwise `false`.
+	cached bool
+	// extractDuration is the amount of time taken to perform extraction, zero if no extraction took place.
+	extractDuration float64
 }
 
-func (rl *Release) Release() {
+// Release is called to signal that the caller has finished using the relase and that resources may be released.
+func (rl *Release) Release(ctx context.Context) {
 	if err := os.Remove(rl.Path); err != nil {
 		logrus.New().WithError(err).Errorf("Failed to delete release link %s", rl.Path)
 	}
+	rl.sendReleaseMetricEvent(ctx)
+}
+
+func (rl *Release) sendReleaseMetricEvent(ctx context.Context) {
+	if !rl.enableInstallerCacheEvent {
+		return
+	}
+	rl.eventsHandler.V2AddMetricsEvent(
+		ctx, &rl.clusterID, nil, nil, "", models.EventSeverityInfo,
+		metricEventInstallerCacheRelease,
+		time.Now(),
+		"release-id", rl.releaseID,
+		"start-time", rl.startTime.Format(time.RFC3339),
+		"cached", rl.cached,
+		"extract-duration", rl.extractDuration,
+	)
 }
 
 // New constructs an installer cache with a given storage capacity
-func New(cacheDir string, storageCapacity int64, log logrus.FieldLogger) *Installers {
+func New(cacheDir string, storageCapacity int64, eventsHandler eventsapi.Handler, enableInstallerCacheEvent bool, log logrus.FieldLogger) *Installers {
 	return &Installers{
-		log:             log,
-		storageCapacity: storageCapacity,
-		cacheDir:        cacheDir,
+		log:                       log,
+		storageCapacity:           storageCapacity,
+		cacheDir:                  cacheDir,
+		eventsHandler:             eventsHandler,
+		enableInstallerCacheEvent: enableInstallerCacheEvent,
 	}
 }
 
 // Get returns the path to an openshift-baremetal-install binary extracted from
 // the referenced release image. Tries the mirror release image first if it's set. It is safe for concurrent use. A cache of
 // binaries is maintained to reduce re-downloading of the same release.
-func (i *Installers) Get(releaseID, releaseIDMirror, pullSecret string, ocRelease oc.Release, ocpVersion string) (*Release, error) {
+func (i *Installers) Get(ctx context.Context, releaseID, releaseIDMirror, pullSecret string, ocRelease oc.Release, ocpVersion string, clusterID strfmt.UUID) (*Release, error) {
 	i.Lock()
 	defer i.Unlock()
 
+	release := &Release{
+		eventsHandler:             i.eventsHandler,
+		enableInstallerCacheEvent: i.enableInstallerCacheEvent,
+		clusterID:                 clusterID,
+		releaseID:                 releaseID,
+		startTime:                 time.Now(),
+	}
+
 	var workdir, binary, path string
 	var err error
 
 	workdir, binary, path, err = ocRelease.GetReleaseBinaryPath(releaseID, i.cacheDir, ocpVersion)
 	if err != nil {
+		release.sendReleaseMetricEvent(ctx)
```

Comment:
I don't think sending an event with incomplete data is a good idea.

This event for example will say that the release wasn't cached but we don't really know that at this point.

I don't feel like we have enough failures in releases to gain anything meaningful from these and even then they might skew the data.

If you really think there's value in these error case events lets create a way to filter them out maybe add an error bool or something :shrug: